jax/interpreters/pxla.py

```diff
@@ -461,13 +474,44 @@ def dynamic_fun(dummy, *args):
     return lambda *_: results
   else:
     compiled, nrep = compile_replicated(jaxpr, backend, axis_name, axis_size,
-                                        devices, consts, *avals)
+                                        global_axis_size, devices, consts,
+                                        *avals)
     device_ordinals = compiled.DeviceOrdinals()
     assignments = assign_shards_to_replicas(nrep, axis_size)
     handle_args = partial(shard_args, backend, device_ordinals, assignments, axis_size)
     handle_outs = _pvals_to_results_handler(axis_size, nrep, out_pvals)
     return partial(execute_replicated, compiled, backend, nrep, handle_args, handle_outs)
 
+def _is_multihost_computation(devices):
+  """Returns if devices specify a multihost computation. devices can be None."""
+  if devices is None:
+    # Default behavior is running on all devices/hosts.
+    return xb.host_count() > 1
+  else:
+    return len(set(d.host_id for d in devices)) > 1
+
+_get_global_axis_size_pmapped = None
+
+def _get_global_axis_size(local_axis_size):
+  """Uses pmap to sum `local_axis_size` across all hosts.
```

Comment:
While this approach is cunning, don't we know the global topology from the low-level runtime stack? Can't we use that here instead?

In particular the module dependency structure seems a bit odd to me with this design.

It also feels strange to me that we have to special case multihost computations here; it feels to me like the same code should work in both single host and multihost cases.

Reply:
As discussed offline, I got rid of this for now and instead require that multihost pmaps run on all devices or have `devices` specified.